### PR TITLE
Implement PaC GitLab integration

### PR DESCRIPTION
**Pull request by mmorhun**
Signed-off-by: Mykola Morhun <mmorhun@redhat.com>

Note: currently, Application Service fails to download devfile from GitLab. To workaround this, **URL to the devfile** must be explicitly set in the Component CR.

Example:
```yaml
apiVersion: appstudio.redhat.com/v1alpha1
kind: Component
metadata:
  annotations:
    pipelinesascode: "1"
  generation: 1
  name: devfile-sample-go-basic
  namespace: default
spec:
  application: test-application
  componentName: devfile-sample-go-basic
  containerImage: quay.io/username/devfile-sample-go-basic
  resources: {}
  source:
    git:
      url: https://gitlab.com/username/devfile-sample-go-basic
      devfileUrl:  https://gitlab.com/username/devfile-sample-go-basic/-/raw/main/devfile.yaml?inline=false
```

### CODE REVIEW

File: controllers/component_build_controller.go [hunk 1]
```diff
@@ -543,8 +549,37 @@ func ConfigureRepositoryForPaC(component appstudiov1alpha1.Component, config map
...
+		}
+
+		gitlabNamespace := gitSourceUrlParts[3]
+		gitlabProjectName := gitSourceUrlParts[4]
+		projectPath := gitlabNamespace + "/" + gitlabProjectName
+
+		err = gitlab.SetupPaCWebhook(glclient, projectPath, webhookTargetUrl, webhookSecret)
+		if err != nil {
+			return "", err
+		}
```
Comment (gabemontero) [inline controllers/component_build_controller.go:564]: I'm not concerned about testing for errors when getting the client, but is there any validation of the inputs into `gitlab.SetupPaCWebhook(glclient, projectPath, webhookTargetUrl, webhookSecret)` that make sense (and maybe add unit tests around those?

Comment (gabemontero) [reply controllers/component_build_controller.go:564]: where we might provide betterr error information to the user than the libraries we are calling?

File: controllers/component_build_controller.go [hunk 2]
```diff
@@ -543,8 +549,37 @@ func ConfigureRepositoryForPaC(component appstudiov1alpha1.Component, config map
...
+			MrTitle:       mrTitle,
+			MrText:        mrText,
+			AuthorName:    authorName,
+			AuthorEmail:   authorEmail,
+			Files: []gitlab.File{
+				{FullPath: ".tekton/" + component.Name + "-" + PipelineRunOnPushFilename, Content: pipelineOnPush},
+				{FullPath: ".tekton/" + component.Name + "-" + PipelineRunOnPRFilename, Content: pipelineOnPR},
+			},
+		}
+		mrUrl, err := gitlab.EnsurePaCMergeRequest(glclient, mrData)
```
Comment (gabemontero) [inline controllers/component_build_controller.go:580]: similar validation of input question here

**Review by gabemontero — COMMENTED** [review]
a couple of minor pre validation before calling dependencies @mmorhun which I are outside the http response code stuff we previously discussed

I do not consider them blockers per se, and full disclosure I did not research the implementations of the github/githlab related libs we are now calling to see if problems inputs are sufficiently described.

File: pkg/gitlab/gitlab_client.go [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+			return false, nil
+		}
+		if !bytes.Equal(fileContent, file.Content) {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func (c *GitlabClient) commitFilesIntoBranch(projectPath, branchName, baseBranch, commitMessage, authorName, authorEmail string, files []File) error {
```
Comment (brunoapimentel) [inline pkg/gitlab/gitlab_client.go:88]: What is the variable `baseBranch` being used for here?

File: controllers/component_build_controller.go (hunk 1 above)
Comment (mmorhun) [reply controllers/component_build_controller.go:564]: > is there any validation of the inputs into gitlab.SetupPaCWebhook(glclient, projectPath, webhookTargetUrl, webhookSecret)

`glclient` is created above. The only argument it gets is the access token.
`projectPath` is built from the repository url. If the url is wrong we'll get 404 later.
`webhookTargetUrl` is read from the Openshift cluster
`webhookSecret` is generated by our code

To sum up the above: we might want to have a better validation of user data (token and git url specifically) by querying something from the project just to check if everything works. What do you think?
Anyway, I'd like to leave that implementation for later, if any.

> where we might provide betterr error information to the user than the libraries we are calling?

I am not sure I got the question correctly, but my plan for the future is to have wrapper/custom error and just log/create event for it.

File: pkg/gitlab/gitlab_client.go (hunk 3 above)
Comment (mmorhun) [reply pkg/gitlab/gitlab_client.go:88]: Leftovers after changing the implementations. Will delete that parameter.

File: controllers/component_build_controller.go (hunk 2 above)
Comment (mmorhun) [reply controllers/component_build_controller.go:580]: Merge request data struct is filled by the operator based on the component fields (name) and git repository url (user name, repository name). If check for git repository url correctness is added, then we are 100% safe here.

Comment (mmorhun) [reply controllers/component_build_controller.go:580]: Moreover, if the git repository url is invalid, I think application service will fail first (on detecting devfile), so build service will not start it's work.
However, I'd still have the check as the logic could be changed one day.

**Comment by brunoapimentel** [comment]
Should we add a test scenario that covers the use of the Gitlab client/helper (similar to the ones we currently have covering Github integration)?

**Review by Michkov — APPROVED** [review]

File: controllers/component_build_controller.go (hunk 1 above)
Comment (gabemontero) [reply controllers/component_build_controller.go:564]: > > is there any validation of the inputs into gitlab.SetupPaCWebhook(glclient, projectPath, webhookTargetUrl, webhookSecret)
> 
> `glclient` is created above. The only argument it gets is the access token. `projectPath` is built from the repository url. If the url is wrong we'll get 404 later. `webhookTargetUrl` is read from the Openshift cluster `webhookSecret` is generated by our code
> 
> To sum up the above: we might want to have a better validation of user data (token and git url specifically) by querying something from the project just to check if everything works. What do you think? Anyway, I'd like to leave that implementation for later, if any.

yep for what it is worth openshift builds does some git ls-remote calls (much less expensive then git clone) to validate git related inputs and catch errors as soon as possible, so there is precedence for such a validation, where that validation is proven useful

I'm fine if you want to leave that for later.  Either add it to your existing jira task around error handling, or open a similar but separate item.

Feel free to mark this item resolved once you've captured things in a Jira.

> 
> > where we might provide betterr error information to the user than the libraries we are calling?
> 
> I am not sure I got the question correctly, but my plan for the future is to have wrapper/custom error and just log/create event for it.

that sounds ^^ exactly around the path I was envisioning

File: controllers/component_build_controller.go (hunk 2 above)
Comment (gabemontero) [reply controllers/component_build_controller.go:580]: sounds good ... feel free to resolve based on your updates per the comment thread above

**Pull request merged**